les/distributor.go

```diff
@@ -0,0 +1,333 @@
+// - canSend tells if the server peer is suitable to serve the request
+// - request prepares sending the request to the given peer and returns a function that
+// does the actual sending. Request order should be preserved but the callback itself should not
+// block until it is sent because other peers might still be able to receive requests while
+// one of them is blocking. Instead, the returned function is put in the peer's orderedSend queue.
+type distReq struct {
+	getCost func(distPeer) uint64
+	canSend func(distPeer) bool
+	request func(distPeer) func()
+
+	reqOrder uint64
+	// only for queued requests
+	noPeersTime          mclock.AbsTime
+	queuePrev, queueNext *distReq
+	sentChn              chan distPeer
+}
+
+// newRequestDistributor creates a new request distributor
+func newRequestDistributor(getAllPeers func() map[distPeer]struct{}, stopChn chan struct{}) *requestDistributor {
+	r := &requestDistributor{
+		loopChn:     make(chan bool, 2),
+		stopChn:     stopChn,
+		getAllPeers: getAllPeers,
+	}
+	go r.loop()
+	return r
+}
+
+// newDistReq creates a new request instance that can be queued for distribution
+func newDistReq(getCost func(distPeer) uint64, canSend func(distPeer) bool, request func(distPeer) func()) *distReq {
+	return &distReq{
+		getCost: getCost,
+		canSend: canSend,
+		request: request,
+	}
+}
+
+// distMaxWait is the maximum waiting time after which further necessary waiting
+// times are recalculated based on new feedback from the servers
+const distMaxWait = time.Millisecond * 10
+
+// main event loop
+func (d *requestDistributor) loop() {
+mainLoop:
+	for {
+		select {
+		case <-time.After(time.Second * 20):
+			fmt.Println("running", d.loopNextSent, "wakeup", d.loopRetrySent)
+			r := d.queueFirst
+			for r != nil {
+				var p uint64
+				if r.queuePrev != nil {
+					p = r.queuePrev.reqOrder
+				}
+				fmt.Println(" order", r.reqOrder, "prev", p)
+				r = r.queueNext
+			}
+		case <-d.stopChn:
+			d.lock.Lock()
+			req := d.queueFirst
+			for req != nil {
+				close(req.sentChn)
+				req = req.queueNext
+			}
+			d.lock.Unlock()
+			return
+		case retry := <-d.loopChn:
+			// the main loop distinguishes two kinds of signals:
+			// - "retry" is sent to re-check whether previously unsendable requests can now be sent to one of the peers
+			// - "next" is sent when there are more requests to be processed/sent
+			// a signal of the same kind should not be sent again to the loop channel before the previous one has been processed
+			d.lock.Lock()
+			if retry {
+				d.loopRetrySent = false
+				if d.loopNextSent {
+					d.lock.Unlock()
+					continue mainLoop
+				}
+			} else {
+				d.loopNextSent = false
+			}
+		loop:
+			for {
+				peer, req, wait := d.nextRequest()
+				if req != nil && wait == 0 {
+					chn := req.sentChn
+					d.remove(req)
+					send := req.request(peer)
+					if send != nil {
+						peer.orderedSend(send)
+					}
+					chn <- peer
```

Comment:
No, I realized that remove sets sentChn to nil so there was actually a reason for this.